src/compiler/checker.ts

```diff
@@ -16743,6 +16746,65 @@ namespace ts {
             }
         }
 
+        function markDecoratorMedataDataTypeNodeAsReferenced(node: TypeNode): void {
+            const entityNameOrToken = getEntityNameForDecoratoryMetadata(node);
+            if (entityNameOrToken && isEntityName(entityNameOrToken)) {
+                markEntityNameOrEntityExpressionAsReference(entityNameOrToken);
+            }
+        }
+
+        type voidUndefinedNullOrNeverTypeNode = Token<SyntaxKind.VoidKeyword | SyntaxKind.UndefinedKeyword | SyntaxKind.NullKeyword | SyntaxKind.NeverKeyword>;
+
+        function getEntityNameForDecoratoryMetadata(node: TypeNode): EntityName | voidUndefinedNullOrNeverTypeNode {
+            if (node) {
+                switch (node.kind) {
+                    case SyntaxKind.IntersectionType:
+                    case SyntaxKind.UnionType:

```

Comment:
why do we allow unions and intersections any ways. we do not serialize them correctly anyways.